src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -329,166 +329,215 @@ roundRate rate rounding =
+        where
+          (notional, calculationFixings) = calculateNotionalAndFixings p s
+            calculationNotionalFixings fxRateId useAdjustedDatesForDcf periodicSchedule
+          createSimpleRate (c, f, dcf) =
+            when (TimeGte f) $ scale notional $ when (TimeGte f) $ scale (rate * dcf) $
+              when (TimeGte p.adjustedEndDate) $ one currency
+            where
+              rate = getFloatingRate s c p regularRate rateFixingCalendar fixLeg
 
       -- Calculate a floating rate correction claim for a payment period
       -- (corresponding to the "-1" in a perf calculation)
+      -- This will no longer be needed when improved compounding logic is supported in CC
       createCorrectionClaim p =
         scale notional $ when (TimeGte p.adjustedEndDate) $ one currency
         where
-          (notional, _, _) = calculateNotionalAndFixings p
+          (notional, _) = calculateNotionalAndFixings p s calculationNotionalFixings fxRateId
+            useAdjustedDatesForDcf periodicSchedule
 
-      claimAmounts = mconcat $ map createClaim paymentSchedule
-      claims = if issuerPaysLeg then claimAmounts else give claimAmounts
-      correctionClaimAmounts = mconcat $ map createCorrectionClaim paymentSchedule
-      correctionClaims = if issuerPaysLeg then give correctionClaimAmounts else
-        correctionClaimAmounts
-      claimsTagged = prepareAndTagClaims dateToDateClockTime [claims, correctionClaims]
-        "Floating rate payment"
-      allTaggedClaims = case s.principalExchanges of
-        None -> [claimsTagged]
-        Some principalExchanges -> [claimsTagged, principalClaimsTagged]
+      -- calculate floating rate spread claim for a payment period
+      createSpreadClaim p =
+        mconcat $ simpleInterestSpreadClaims
+        where
+          (notional, calculationFixings) = calculateNotionalAndFixings p s
+            calculationNotionalFixings fxRateId useAdjustedDatesForDcf periodicSchedule
+          spread = Const floatingRateSpread
+          simpleInterestSpreadClaims = map
+            (\(c, f, dcf) -> scale notional $ scale spread $ scale dcf $
+              when (TimeGte p.adjustedEndDate) $ one currency)
+            (reverse calculationFixings)
+
+      calculateStraightCompoundingClaims =
+        [claims, correctionClaims]
           where
-            principalClaimsTagged = calculatePrincipalExchangePaymentClaims paymentSchedule
-              issuerPaysLeg currency fxRateId notionals rateFixingDates principalExchanges
+            claimAmounts = mconcat $ map (createClaim regularRateInclSpread) paymentSchedule
+            claims = if issuerPaysLeg then claimAmounts else give claimAmounts
+            correctionClaimAmounts = mconcat $ map createCorrectionClaim paymentSchedule
+            correctionClaims = if issuerPaysLeg then give correctionClaimAmounts else
+              correctionClaimAmounts
+
+      calculateSpredExclusiveCompoundingClaims =
```

Comment:
What do you think would be a good function signature for a generic compounding helper function?
I think the current split is quite descriptive for rates compounding:
```
          compoundRates acc (c, f, dcf) =
            when (TimeGte f) $ scale (Const 1.0 + rate * dcf) acc
```
but to make it generic I wonder if one should e.g. combine `rate * dcf` into a rateUsed. That would require a couple of (probably minor) changes to the current calculation / data aggregation logic.

Reply:
Maybe better to do this as part of a separate PR, as this one now has lots of changes.

Reply:
I think the signature of this function is quite nice. 

I guess I was referring to the broader calculation, where you decide which compounding method to use and then apply it

Reply:
I see. I have removed some of those functions now and put the logic directly the case statement, since it is now less code when I only do `if issuerPaysLeg then claimAmounts else fmap give claimAmounts` once.

Reply:
Do you think it is worth moving out the others? I find that that it is maybe a bit overkill, since they are just used within this function and they would then get more arguments which would make the `map` statements more difficult to read.

Reply:
fine for me to keep as is